### PR TITLE
[real-time-transcription] Provider-neutral async Soniox [S01-W02-P01] + plan stage S05

### PR DESCRIPTION
**Pull request by AlexDochioiu**
## Summary

Implements plan step **S01-W02-P01** (`.plan/active/real-time-transcription/stages/01-data-and-async/w02-p01-provider-neutral-async-soniox.md`), plus a plan-document change adding **stage S05**.

**Async transcription is now provider-neutral.** `VoiceService` depends on one `AsyncTranscriptionClient` interface with two implementations; OpenAI remains selected by default and its behavior is unchanged.

- `AsyncTranscriptionClient` + a closed provider-neutral failure enum
- OpenAI adapted in place, keeping its shipped HTTP 500 mapping behind a `COMPATIBILITY` marker
- Soniox client behind a narrow structural SDK type, with every consumed SDK value validated through Zod adapters
- deploy-time provider selection, EU-pinned endpoints, conditional key requirement
- immediate cleanup of provider audio plus an operator audit/purge command
- Soniox disclosed as an EU sub-processor in the privacy policy

**Plan stage S05** was added because the plan mandated *retaining* glossary rollback tooling and several `COMPATIBILITY` markers but never scheduled their deletion — the debt was orphaned. S05 carries a per-item trigger inventory and an explicit retention list. Notably, legacy index *detection* and the startup guard are permanent; only *reversal* paths are in scope, since deleting detection would blind the guard shipped in #61.

## Three defects found and fixed during review

Each was reproduced against the real SDK before fixing, and each fix was mutation-verified by reintroducing the defect and confirming the tests fail.

1. **EU residency pin failed open.** Passing `region: "eu"` does *not* pin the endpoint — the SDK resolves `options.base_url ?? process.env.SONIOX_API_BASE_URL ?? regionDefaults`. Confirmed: with `SONIOX_API_BASE_URL` set, a `region:"eu"` client resolved to the attacker host, so audio and the API key would have been sent there. Now pinned via `base_url`, with a test proving it against a hostile env var.
2. **Every Soniox request would have failed.** `stt.create` returns a `queued`/`processing` job, but it was validated with a completed-only parser — so each request returned HTTP 504 *and* stranded a transcription at Soniox, manufacturing exactly the residue the purge command exists to clean. The test fixture returned `status: "completed"`, a shape the SDK never emits, which hid it.
3. **Cancellation and timeout were unreachable.** The SDK reports aborts as `SonioxHttpError`/`code: "aborted"`, never a DOM `AbortError`, so both fell through to HTTP 503. Cancellation and timeout are now decided from the signals we own, matching what the OpenAI adapter already did.

All three were the same class: trusting an assumed SDK shape instead of the one the provider actually emits. Fixtures now bind the real contract.

## Verification

MongoDB backend disclosure: run against the `mongodb-memory-server` fallback (real `mongod`), **not** MongoDB 7, since `MONGODB_URI_TEST` was unset locally. No schema, collection, index, or persistence contract changed. CI enforces MongoDB 7.

- full suite: 621 tests, 620 passed, 1 pre-existing skip, 0 failed
- new: `tests/api/soniox-transcription-api.test.ts`, `tests/clients/async-transcription-clients.test.ts`, `tests/services/voice-service.test.ts` (route-level assertions for every failure reason, incl. exact status / `retryable` / `Retry-After` presence and the quota-vs-capacity distinction), `tests/scripts/purge-soniox-transcription.test.ts`
- `format:check`, `lint`, `tsc --noEmit`, `build`, `circular-dependencies` all pass
- `docker build` passes with no provider credentials and no provider egress
- authored/generated diff: 2,051 / 7 (lockfile)

No real-Soniox evidence is recorded — no credential exists locally, and the step file states CI requires none.

## Rollout gates (do not skip)

Soniox is disabled by default and **must not be selected in production** until: the DPA is signed, the dedicated EU project is provisioned, the updated privacy policy is published, and the mobile app's request timeout exceeds the server budget. Soniox async can exceed the app's current 30-second timeout, which is why S03 ships before any provider switch.

## Next slice

S02-W01-P01 — realtime auth proxy, registered behind a default-false flag.

---
## Summary by cubic
Makes async transcription provider‑neutral and adds a validated, EU‑pinned Soniox async client; OpenAI remains the default. Centralizes Retry‑After handling and hardens error logs by traversing provider cause chains and redacting credentials.

- Introduces `AsyncTranscriptionClient` with a closed failure enum; the global handler is the only writer of `Retry-After`, honoring provider cooldowns (clamped ≤300s) with a 5s fallback for unquantified capacity.
- Adds Soniox via `@soniox/node`: Zod‑validated values, explicit EU REST `base_url` (ignores env precedence), local-signal decisions for cancel vs timeout, job‑ID correlation, and per‑phase cleanup (delete transcription, then file) with independent budgets. Includes a `purge-soniox-transcription` audit/apply command with bounded delete concurrency, per‑item timeouts, a file‑sweep holdback on transcription uncertainty, and flush of partial batches on list failures.
- Behavior changes: the voice route cancels the provider request if the client disconnects; unusable audio returns 400; empty/whitespace transcripts and missing/non‑positive/>24h durations fail as malformed output and do not consume quota; provider messages are never exposed to clients.
- Settings: `DELETE /auth/settings` clears all device overrides for the authenticated account; idempotent and rate‑limited 30/min independently from PATCH.
- Composition/docs/plan: deploy‑time provider selection (`ASYNC_TRANSCRIPTION_PROVIDER`), OpenAI remains default; privacy policy discloses Soniox; adds plan stage S05 to schedule removal of compatibility scaffolding.

**Rollout gates**
- Soniox stays disabled by default. Do not enable in production until: DPA signed, dedicated EU project provisioned, updated privacy policy published, and the mobile app timeout exceeds the server budget.

<sup>Written for commit 10ae85df45c761d20b5e96266970b51266ecd9cb. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/sesori-ai/sesori_auth_server/pull/64?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: src/api/soniox-transcription-api.ts [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+ * partial second is never billed as zero, matching the OpenAI estimate policy.
+ * A provider that omits duration is malformed output rather than a free request.
+ */
+export function toBillableSeconds(audioDurationMs: number | null): number {
+  if (audioDurationMs === null) {
+    fail(TranscriptionFailureReason.MalformedOutput);
+  }
+
+  return Math.ceil(audioDurationMs / 1000);
+}
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply src/api/soniox-transcription-api.ts:178]: [Sesori reply] Addressed (in commit eaea7ea)

Correct, and the test you flagged was locking in the wrong behavior. A zero duration is now rejected as `malformed_output` rather than silently billing as free, and `toBillableSeconds` floors at one second so a sub-second clip still costs something. The `toBillableSeconds(0) === 0` assertion is gone, replaced by one asserting null/0/-1 all fail closed.

File: src/scripts/purge-soniox-transcription.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    deletedTranscriptionCount: 0,
+  };
+
+  try {
+    // Transcriptions first: they reference files.
+    for await (const transcription of await input.sdk.stt.list()) {
+      report.transcriptionCount += 1;
+      if (input.mode === "apply") {
+        await input.sdk.stt.delete(transcription.id);
+        report.deletedTranscriptionCount += 1;
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply src/scripts/purge-soniox-transcription.ts]: [Sesori reply] Addressed (in commit eaea7ea)

Each list and delete now runs under its own `AbortSignal.timeout`, so a stalled provider request can no longer hang the maintenance command indefinitely.

File: src/api/soniox-transcription-api.ts [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+ * value flowing into billing or a transcript response.
+ */
+
+const transcriptionStatusSchema = z.enum(["queued", "processing", "completed", "error"]);
+
+const completedTranscriptionSchema = z
+  .object({
+    id: z.string().min(1),
+    status: transcriptionStatusSchema,
+    audio_duration_ms: z.number().finite().nonnegative().nullish(),
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply src/api/soniox-transcription-api.ts]: [Sesori reply] Addressed (in commit eaea7ea)

Valid. `audio_duration_ms` is now `positive()` with a 24-hour ceiling, and `toBillableSeconds` rejects null or non-positive and floors at one second. A zero-duration result previously billed as free; an absurd one could have consumed the whole daily quota in a single request. Both are covered by new tests.

File: src/clients/soniox-transcription-client.ts [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+      await this.#cleanup({ fileId, transcriptionId });
+    }
+  }
+
+  async #cleanup(ids: { fileId: string | null; transcriptionId: string | null }): Promise<void> {
+    const signal = AbortSignal.timeout(this.#cleanupTimeoutMs);
+
+    // Transcription first: it references the file.
+    if (ids.transcriptionId !== null) {
+      await this.#deleteQuietly(() => this.#sdk.stt.delete(ids.transcriptionId as string, signal), "transcription");
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply src/clients/soniox-transcription-client.ts]: [Sesori reply] Partially addressed (in commit eaea7ea)

Took the unsafe-ordering half: `#cleanup` now returns early when the transcription delete fails, so the file is never deleted out from under a job that may still exist. Both are left for the operator purge instead. New test asserts only `stt.delete` is attempted on that path.

Did not add terminal-state tracking or a reconciliation queue. Deleting a still-active job is the intended behavior on the timeout/failure path — the request is over and the work is abandoned, so leaving the job running would strand exactly the residue this cleanup exists to prevent. A durable reconciliation store is explicitly out of scope for this plan (PLAN.md §3 non-goals: no job collection, receipts, or reconciler); the purge command is the sanctioned recovery for what a hard crash leaves behind.

File: src/scripts/purge-soniox-transcription.ts [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+      }
+    }
+
+    for await (const file of await input.sdk.files.list()) {
+      report.fileCount += 1;
+      if (input.mode === "apply") {
+        await input.sdk.files.delete(file.id);
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply src/scripts/purge-soniox-transcription.ts]: [Sesori reply] Partially addressed (in commit eaea7ea)

The risk is real and I have documented it rather than guessed at a staleness rule: `--help` and the README now state that `--apply` must run only while no Soniox traffic is in flight, because it deletes every listed file and job including one an active request just created.

I did not add automatic age-based filtering. The SDK's list payload has no field this code already validates as a reliable creation timestamp, and inventing a staleness heuristic would risk the opposite failure — silently leaving real residue behind while appearing to have cleaned. Audit mode is the default and is read-only, so the destructive path is explicit and operator-gated. If you want an age filter, it should be specified against a confirmed provider field rather than assumed.

File: src/api/soniox-transcription-api.ts [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+    error_type: z.string().nullish(),
+    error_message: z.string().nullish(),
+  })
+  .loose();
+
+const transcriptSchema = z
+  .object({
+    text: z.string(),
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply src/api/soniox-transcription-api.ts]: [Sesori reply] Addressed (in commit eaea7ea)

Transcript text is now bounded at 1,000,000 characters at the adapter boundary, so an oversized provider response fails closed as `malformed_output` rather than reaching the route's late empty-text check. Test added.

Kept empty text accepted at the schema level: the route already converts an empty transcript into a 500, and treating silence as malformed provider output would misreport a legitimately silent recording.

File: src/services/voice-service.ts [hunk 7]
```diff
@@ -50,20 +82,47 @@ export class VoiceService {
...
     return { text, dailySecondsRemaining };
   }
+
+  /** Maps the closed provider-neutral failure enum onto the public API contract. */
+  static #toApiError(error: unknown): Error {
+    if (!(error instanceof TranscriptionFailure)) {
+      return new InternalServerError({ debugMessage: "Transcription failed", nestedError: error });
+    }
+
+    switch (error.reason) {
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply src/services/voice-service.ts:133]: [Sesori reply] Addressed (in commit eaea7ea)

`error.cause` is now passed as `nestedError` on every mapped `ApiError`. The global handler logs it server-side and never includes it in the response body, so diagnosis improves without widening the public contract.

File: src/services/voice-service.ts [hunk 8]
```diff
@@ -50,20 +82,47 @@ export class VoiceService {
...
+      case TranscriptionFailureReason.InvalidInput:
+        return new BadRequestError({ debugMessage: "Provider rejected the audio input" });
+      case TranscriptionFailureReason.Capacity:
+      case TranscriptionFailureReason.Unavailable:
+        return new TranscriptionUnavailableError({ debugMessage: "Transcription provider unavailable" });
+      case TranscriptionFailureReason.Timeout:
+        return new TranscriptionTimeoutError({ debugMessage: "Transcription provider timed out" });
+      case TranscriptionFailureReason.ProviderRejected:
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply src/services/voice-service.ts:157]: [Sesori reply] Addressed (in commit eaea7ea)

Good catch — `bad_audio` was reported as a 500 configuration fault, blaming the operator for a caller's unusable upload.

`parseTranscription` now inspects `error_type`: a known audio-rejection value maps to `invalid_input` → 400, while anything else, including an absent type, stays `provider_rejected` → 500. Unknown values deliberately keep the configuration mapping, since it is safer not to blame the caller for a failure we cannot classify. Tests cover both directions.

File: src/clients/soniox-transcription-client.ts [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+      throw new TranscriptionFailure(reason, { cause: error });
+    } finally {
+      // Cleanup gets its own budget so a caller abort or request timeout cannot
+      // strand provider-side audio. Failure is logged, never surfaced.
+      await this.#cleanup({ fileId, transcriptionId });
+    }
+  }
+
+  async #cleanup(ids: { fileId: string | null; transcriptionId: string | null }): Promise<void> {
+    const signal = AbortSignal.timeout(this.#cleanupTimeoutMs);
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply src/clients/soniox-transcription-client.ts]: [Sesori reply] Addressed (in commit eaea7ea)

Each delete now gets its own `AbortSignal.timeout(cleanupTimeoutMs)` instead of sharing one budget, so a slow job delete can no longer consume the file delete's time and strand uploaded audio.

File: src/scripts/purge-soniox-transcription.ts [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    deletedFileCount: 0,
+    deletedTranscriptionCount: 0,
+  };
+
+  try {
+    // Transcriptions first: they reference files.
+    for await (const transcription of await input.sdk.stt.list()) {
+      report.transcriptionCount += 1;
+      if (input.mode === "apply") {
+        await input.sdk.stt.delete(transcription.id);
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply src/scripts/purge-soniox-transcription.ts]: [Sesori reply] Addressed (in commit eaea7ea)

Every listed item is now validated with a Zod schema before its `id` reaches a delete call, consistent with the request-path adapters. A malformed item is counted as a failure and skipped rather than sent to the endpoint. Test added.

File: src/scripts/purge-soniox-transcription.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply src/scripts/purge-soniox-transcription.ts]: [Sesori reply] Addressed (in commit eaea7ea)

Every delete now runs under `AbortSignal.timeout`, defaulting to the same 10s budget as request-path cleanup and injectable for tests.

File: src/scripts/purge-soniox-transcription.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply src/scripts/purge-soniox-transcription.ts]: [Sesori reply] Addressed (in commit eaea7ea)

Failures are now isolated per resource: each delete is wrapped individually, the sweep continues, and the command reports `failed` at the end. Previously one failed delete abandoned every later transcription and all files. New test asserts all three resources are attempted when the first delete throws, with `outcome: "failed"`.

File: tests/api/soniox-transcription-api.test.ts [hunk 11]
```diff
@@ -0,0 +1,151 @@
+import { describe, it } from "node:test";
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply tests/api/soniox-transcription-api.test.ts:1]: [Sesori reply] Addressed (in commit eaea7ea)

Fair — and its contract is exactly the one an earlier review round found broken, so it deserved direct coverage. Added tests asserting it accepts `queued` and `processing` (the statuses creation actually returns) and rejects a missing, empty, or non-string id.

File: src/clients/openai-client.ts [hunk 12]
```diff
@@ -15,26 +22,41 @@ export class OpenAIClient {
...
-      this.#client.audio.transcriptions.create({
-        file,
-        model: this.#model,
-        language: "en",
-        response_format: "json",
-        ...(args.prompt ? { prompt: args.prompt } : {}),
-      }),
-      OpenAIClient.parseAudioDuration(args.fileBuffer, args.mimetype),
-    ]);
+  private static buildPrompt(terms: string[]): string | null {
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply src/clients/openai-client.ts]: [Sesori reply] Addressed (in commit eaea7ea)

You caught a real regression I introduced in this PR: moving prompt rendering into the adapter left `GlossaryService.buildTranscriptionPrompt` reachable only from its own tests, while the two copies of the wrapper could drift — and `getContextWords` was reserving budget using its private copy.

Both now use one exported `renderTranscriptionPrompt` plus shared prefix/suffix constants. The dead service method is deleted and its tests retargeted at the shared renderer, so the reserved budget and the rendered string cannot disagree.

File: src/api/soniox-transcription-api.ts [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+
+const transcriptionStatusSchema = z.enum(["queued", "processing", "completed", "error"]);
+
+const completedTranscriptionSchema = z
+  .object({
+    id: z.string().min(1),
+    status: transcriptionStatusSchema,
+    // Bounded: a non-positive duration would bill as free and an absurd one
+    // would blow the daily quota. 24h is far above any supported upload.
+    audio_duration_ms: z.number().finite().positive().max(86_400_000).nullish(),
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply src/api/soniox-transcription-api.ts]: [Sesori reply] Addressed (in commit f7aac75)

You're right, and this was a regression from my previous commit — I verified it: `{ status: "error", error_type: "bad_audio", audio_duration_ms: 0 }` classified as `malformed_output` (502) instead of `invalid_input` (400), because the schema short-circuited before the `error_type` logic ran.

The duration bound now applies only after the status is known to be `completed`, where it is the only place duration is actually billable. A failed job's duration is meaningless and no longer influences classification. Test added covering both `bad_audio` and a credential-style error with bad durations.

File: src/api/soniox-transcription-api.ts [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+    error_type: z.string().nullish(),
+    error_message: z.string().nullish(),
+  })
+  .loose();
+
+const maxTranscriptCharacters = 1_000_000;
+
+const transcriptSchema = z
+  .object({
+    text: z.string().max(maxTranscriptCharacters),
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply src/api/soniox-transcription-api.ts]: [Sesori reply] Addressed (in commit f7aac75)

Agreed on reflection — I previously argued empty text should pass so a silent recording wasn't misreported, but that reasoning was wrong for the billing consequence you point out: the request consumed quota and then failed as a generic 500.

`parseTranscriptText` now rejects empty and whitespace-only text as `malformed_output`, before usage is recorded. Test covers `""`, spaces, and tabs/newlines.

File: src/scripts/purge-soniox-transcription.ts [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+      const parsed = purgeItemSchema.safeParse(item);
+      if (!parsed.success) {
+        markFailed("file_item", parsed.error.issues);
+        continue;
+      }
+
+      report.fileCount += 1;
+      const removed =
+        input.mode === "apply" &&
+        (await deleteOne((signal) => input.sdk.files.delete(parsed.data.id, signal), "file_delete"));
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply src/scripts/purge-soniox-transcription.ts]: [Sesori reply] Addressed (in commit f7aac75)

Correct, and thank you for catching the inconsistency — the purge was violating the exact rule I had just added to the client cleanup.

The file sweep is now held back whenever any transcription delete failed, any transcription list item was malformed, or the transcription listing itself failed. Files are still counted and reported, just not deleted, so the operator sees the residue without the command creating a job that references missing audio. An existing test that asserted the old behavior was updated to the corrected contract, and a new test covers the held-back sweep.

File: src/api/soniox-transcription-api.ts [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+    fail(TranscriptionFailureReason.Timeout);
+  }
+
+  // Only a completed job's duration is billable, so enforce the bound here:
+  // non-positive would bill as free, absurd would blow the daily quota.
+  if (audioDurationMs !== null && audioDurationMs !== undefined) {
+    if (audioDurationMs <= 0 || audioDurationMs > maxAudioDurationMs) {
+      fail(TranscriptionFailureReason.MalformedOutput);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply src/api/soniox-transcription-api.ts]: [Sesori reply] Addressed (in commit 0a46e22)

Fair tightening. A completed job with no duration was reaching the client and only failing later in `toBillableSeconds`, so a transcript-fetch failure in between could surface it as a different error than the malformed output it is.

`parseTranscription` now rejects absent durations alongside non-positive and over-24-hour values in the completed branch. `ValidatedTranscription.audioDurationMs` is consequently non-nullable, which removes the nullable handling from callers. Test covers both the missing and explicit-null cases.

File: src/scripts/purge-soniox-transcription.ts [hunk 17]
```diff
@@ -0,0 +1,254 @@
...
+  // Transcriptions first: they reference files.
+  try {
+    let pending: string[] = [];
+    const flush = async (): Promise<void> => {
+      if (pending.length === 0) {
+        return;
+      }
+
+      const ids = pending;
+      pending = [];
```
[the remark this answers is not in the document]
Comment (stanvladut) [reply src/scripts/purge-soniox-transcription.ts]: [Sesori reply] Addressed (in commit 8ba2148)

The pending IDs the iterator already yielded are now retained across the throw: the batch buffer lives outside the try/catch and the flush runs after the loop whether it completed or threw. That preserves best-effort cleanup while the sweep is still marked failed, and the transcription-uncertainty rule keeps holding files back. Non-Error SDK throws are also wrapped in a local `TypeError` with the original as `cause` before reaching `markFailed`, so the count-only log line can't inherit a provider string. Two focused regression subtests in `tests/scripts/purge-soniox-transcription-list-failure.test.ts` fail on the prior implementation and pass now.

File: src/api/soniox-transcription-api.ts [hunk 18]
```diff
@@ -0,0 +1,307 @@
...
+  // A completed job must carry a billable duration. Rejecting absent, non-
+  // positive, and absurd values here classifies malformed provider output at
+  // the adapter boundary rather than leaving it to a later billing step.
+  // The upper bound is exclusive: a full 24h async job is itself out of
+  // contract, so equality is rejected rather than billed.
+  if (
+    audioDurationMs === null ||
+    audioDurationMs === undefined ||
+    audioDurationMs <= 0 ||
+    audioDurationMs >= maxAudioDurationMs
```
[the remark this answers is not in the document]
Comment (stanvladut) [reply src/api/soniox-transcription-api.ts]: [Sesori reply] Addressed (in commit 6764ea8)

The completed-job duration check is back to `> maxAudioDurationMs`, matching the provider contract, and the paired test now pins both edges: 86_399_999 ms still passes and 86_400_000 ms is accepted rather than rejected. Absent, non-positive, and above-boundary values still fail as `malformed_output`.

File: tests/legal/routes.test.ts [hunk 19]
```diff
@@ -89,4 +93,31 @@ describe("Legal routes", () => {
...
     }
   });
+
+  it("serves the real privacy asset disclosing every AI sub-processor", async (t) => {
+    // The other route tests inject synthetic text, so nothing else proves the
+    // shipped asset actually discloses the providers we send audio to.
+    mockMongoHarness(t);
+    // Resolved from the composition root's module URL, the way production does,
+    // rather than from this test file's own directory.
+    const compositionRootUrl = pathToFileURL(path.join(process.cwd(), "src/index.ts")).href;
```
[the remark this answers is not in the document]
Comment (stanvladut) [reply tests/legal/routes.test.ts]: [Sesori reply] Addressed (in commit 111b9af)

The composition-root URL is now built with `new URL("../../src/index.ts", import.meta.url).href`, so the test resolves the shipped legal assets independently of the caller's cwd. The obsolete `path` and `pathToFileURL` imports are dropped along with the `process.cwd()` reference.

File: tests/clients/async-transcription-clients.test.ts [hunk 20]
```diff
@@ -0,0 +1,513 @@
...
+  }
+
+  it("returns text and a positive billable duration", async (t) => {
+    t.mock.method(Transcriptions.prototype, "create", async () => ({ text: "hello from openai" }));
+
+    const result = await createOpenAIClient().transcribe(request());
+
+    assert.equal(result.text, "hello from openai");
+    assert.ok(Number.isInteger(result.durationSeconds), "duration must be whole seconds");
+    assert.ok(result.durationSeconds > 0, "a sub-second clip must never bill as free");
```
[the remark this answers is not in the document]
Comment (stanvladut) [reply tests/clients/async-transcription-clients.test.ts]: [Sesori reply] Addressed (in commit 6c50a3c)

The OpenAI adapter now has its own suite in `tests/clients/openai-transcription-client.test.ts` and drives the `music-metadata` path end-to-end with a real 0.2s PCM WAV. `durationSeconds` is asserted against the parsed duration to millisecond precision, so the assertion no longer coincidentally passes only because the size-estimate fallback rounds up. A second test keeps the 10-byte fallback pinned at 1 s so that path stays regression-covered. `AsyncTranscriptionClient.transcribe` documentation is updated in the same commit to say implementations may preserve fractional metadata precision, so consumers must not assume whole seconds. The `malformed_output` contract for absent/non-positive/absurd durations and the OpenAI COMPATIBILITY collapse-to-`internal` behaviour are unchanged and still asserted.

File: src/clients/soniox-transcription-client.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (stanvladut) [reply src/clients/soniox-transcription-client.ts]: [Sesori reply] Partially addressed

No further code change in this round. This thread already received a partial fix in commit eaea7ea: cleanup no longer deletes the file if transcription deletion fails, preserving the safe deletion order. I am leaving terminal-state tracking/reconciliation out because the plan explicitly uses the purge command for hard-crash residue and does not introduce a durable job collection/reconciler in this slice.

File: src/scripts/purge-soniox-transcription.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (stanvladut) [reply src/scripts/purge-soniox-transcription.ts]: [Sesori reply] Partially addressed

No further code change in this round. This thread already received a partial fix in commit eaea7ea by documenting the required quiescence for destructive purge apply mode in help/README. I am not adding an automatic age filter because the SDK list shape is not currently validated to provide a reliable creation timestamp, and an invented heuristic would make cleanup appear complete while silently leaving residue.

File: src/services/voice-service.ts [hunk 21]
```diff
@@ -50,20 +102,78 @@ export class VoiceService {
...
     return { text, dailySecondsRemaining };
   }
+
+  /** Maps the closed provider-neutral failure enum onto the public API contract. */
+  static #toApiError(error: unknown): Error {
+    if (!(error instanceof TranscriptionFailure)) {
+      return new InternalServerError({ debugMessage: "Transcription failed", nestedError: error });
+    }
+
+    const cause = { errorType: safeErrorType({ error: error.cause }) };
```
[the remark this answers is not in the document]
Comment (stanvladut) [reply src/services/voice-service.ts]: [Sesori reply] Addressed (in commit 278ca6d)

Restored the provider failure cause as the ApiError nestedError so the global error handler logs the provider message/properties for diagnosis, while the HTTP response still contains only the public transcription_configuration_error payload. The regression test now asserts both sides of that contract.

File: tests/services/voice-service.test.ts [hunk 22]
```diff
@@ -0,0 +1,216 @@
...
+      };
+
+      const res = await post(user.accessToken);
+
+      assert.equal(res.statusCode, 500);
+      assert.deepEqual(res.json(), { error: "transcription_configuration_error", retryable: false });
+      const logs = loggedEntries.join("\n");
+      assert.match(logs, /SensitiveProviderError/);
+      assert.match(logs, new RegExp(providerMessageSentinel));
+      assert.match(logs, new RegExp(providerPropertySentinel));
```
[the remark this answers is not in the document]
Comment (stanvladut) [reply tests/services/voice-service.test.ts:170]: [Sesori reply] Addressed (in commit 91ac3dc)

Added provider-cause header redaction before handing nestedError to the global logger. The logger still receives the provider error type, message, and non-sensitive request metadata for diagnosis, but authorization/cookie/proxy-authorization header values are replaced with [redacted]. The route regression now asserts public response hygiene, diagnostic retention, and credential redaction.

File: src/services/voice-service.ts [hunk 23]
```diff
@@ -50,20 +103,112 @@ export class VoiceService {
...
+    return input.cause;
+  }
+
+  const headers = Reflect.get(input.cause, "headers");
+  if (headers === null || typeof headers !== "object") {
+    return input.cause;
+  }
+
+  if (input.cause instanceof Error) {
+    const redactedError = new Error(input.cause.message, { cause: input.cause.cause });
```
[the remark this answers is not in the document]
Comment (stanvladut) [reply src/services/voice-service.ts]: [Sesori reply] Addressed (in commit 99d28b2)

Redacted nested provider causes before attaching them to the copied Error. The regression now includes an authorization header inside Error.cause and asserts that the nested credential is absent from captured server logs.

File: src/services/voice-service.ts [hunk 24]
```diff
@@ -50,20 +103,114 @@ export class VoiceService {
...
+    return input.cause;
+  }
+
+  const headers = Reflect.get(input.cause, "headers");
+  if (headers === null || typeof headers !== "object") {
+    return input.cause;
+  }
+
+  if (input.cause instanceof Error) {
+    const redactedCause = redactProviderCause({ cause: input.cause.cause });
```
[the remark this answers is not in the document]
Comment (stanvladut) [reply src/services/voice-service.ts:188]: [Sesori reply] Addressed (in commit 10ae85d)

Moved Error cause traversal ahead of the header guard so every Error in the provider cause chain is copied with a sanitized nested cause, even when intermediate errors do not have headers. The regression now includes a deeper cookie-bearing cause behind an intermediate wrapper without headers.
